### PR TITLE
feat: Date filters on bank reconciliation tool

### PR DESCRIPTION
**Pull request by sonali8848**
https://user-images.githubusercontent.com/115159164/211467246-7cea9fc2-8a93-4af2-9c84-6f5f0eb35186.mp4

**Date filters in Bank Reconciliation Tool** 
Step 1-To apply date filter as per posting date in Bank Transactions and Vouchers and sorted bank transactions as per ascending order.
Screenshot 2022-12-08 170131
Step 2-  Added checkbox of 'filtered as per reference date' and two fields 'from reference date' and 'to reference date' to filter by reference date. If the checkbox is ticked then the vouchers will be filtered and sorted as per the reference date. 
Screenshot 2022-12-08 172838
Step 3- Added field of posting date in the dialog.
Screenshot 2022-12-08 173810
Step 4- Journal Entry checkbox is ticked by default.
Screenshot 2022-12-08 173810

**Consolidated auto bank reconciliation**

All the transactions will get auto-reconciled if the reference number matched with bank uploaded transaction and voucher posted, we can see that that pop message will show the count of transactions reconciled and if no transactions are reconciled then the pop will show “There is no matched reference number for reconciliation.
image
image

### CODE REVIEW

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 1]
```diff
@@ -25,6 +25,8 @@ class BankReconciliationTool(Document):
 @frappe.whitelist()
 def get_bank_transactions(bank_account, from_date=None, to_date=None):
 	# returns bank transactions for a bank account
+	from_date = frappe.db.get_single_value("Bank Reconciliation Tool", "bank_statement_from_date")
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]: This is the incorrect way of getting the dates, pass from the client side. I can see options params are already there

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 2]
```diff
@@ -51,6 +53,7 @@ def get_bank_transactions(bank_account, from_date=None, to_date=None):
 		],
 		filters=filters,
 	)
+	transactions = sorted(transactions, key=lambda x: x["date"]) if transactions else []
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]: Can add a order by clause in query itself, this seems unnecessary

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 3]
```diff
@@ -340,6 +343,9 @@ def get_linked_payments(bank_transaction_name, document_types=None):
 
 def check_matching(bank_account, company, transaction, document_types):
 	# combine all types of vouchers
+	filtered_by_reference_date = frappe.db.get_single_value(
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]: Pass from client side, don't make a DB call

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 4]
```diff
@@ -506,45 +520,103 @@ def get_lr_matching_query(bank_account, amount_condition, filters):
...
 	# get matching payment entries query
+	from_date = frappe.db.get_single_value("Bank Reconciliation Tool", "bank_statement_from_date")
+	to_date = frappe.db.get_single_value("Bank Reconciliation Tool", "bank_statement_to_date")
+	from_reference_date = frappe.db.get_single_value(
+		"Bank Reconciliation Tool", "from_reference_date"
+	)
+	to_reference_date = frappe.db.get_single_value("Bank Reconciliation Tool", "to_reference_date")
+	filtered_by_reference_date = frappe.db.get_single_value(
+		"Bank Reconciliation Tool", "filtered_by_reference_date"
+	)
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]: Pass from the client side

File: erpnext/hooks.py [hunk 5]
```diff
@@ -469,8 +469,8 @@
 bank_reconciliation_doctypes = [
 	"Payment Entry",
 	"Journal Entry",
-	"Purchase Invoice",
-	"Sales Invoice",
+	# "Purchase Invoice",
```
Comment (deepeshgarg007) [inline erpnext/hooks.py]: Why are they commented?

File: erpnext/public/js/bank_reconciliation_tool/dialog_manager.js [hunk 6]
```diff
@@ -35,13 +35,590 @@ erpnext.accounts.bank_reconciliation.DialogManager = class DialogManager {
...
+					},
+					callback: (r) => {
+						var doc = frappe.model.sync(r.message);
+						frappe.set_route("Form", doc[0].doctype, doc[0].name);
+					},
+				});
+			}
+		}
+	
+	};
```
Comment (deepeshgarg007) [inline erpnext/public/js/bank_reconciliation_tool/dialog_manager.js]: What is this for, can't see any mention in the description. Please be more descriptive

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]
- Assuming the user will be filtering either by reference or posting date. So when filter by reference date is enabled those fields should hide
- The commits are non-semantic, please add semantic comments
- This is the first level review, will add more once this cleanup is done

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js [hunk 7]
```diff
@@ -160,6 +160,9 @@ frappe.ui.form.on("Bank Reconciliation Tool", {
 					).$wrapper,
 					bank_statement_from_date: frm.doc.bank_statement_from_date,
 					bank_statement_to_date: frm.doc.bank_statement_to_date,
+					filtered_by_reference_date:frm.doc.filtered_by_reference_date,
+					from_reference_date:frm.doc.from_reference_date,
+					to_reference_date:frm.doc.to_reference_date,
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js]:
```suggestion
					filtered_by_reference_date: frm.doc.filtered_by_reference_date,
					from_reference_date: frm.doc.from_reference_date,
					to_reference_date: frm.doc.to_reference_date,
```

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 8]
```diff
@@ -504,47 +584,75 @@ def get_lr_matching_query(bank_account, amount_condition, filters):
...
+	to_reference_date,
+):
 	# get matching payment entries query
 	if transaction.deposit > 0:
 		currency_field = "paid_to_account_currency as currency"
 	else:
 		currency_field = "paid_from_account_currency as currency"
+	filter_by_date = f"AND posting_date between '{from_date}' and '{to_date}'"
+	order_by = " posting_date"
+	if filtered_by_reference_date == "1":
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]: Should be an int, why comparing with a string

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 9]
```diff
@@ -504,47 +584,75 @@ def get_lr_matching_query(bank_account, amount_condition, filters):
...
+):
 	# get matching journal entry query
-
 	# We have mapping at the bank level
 	# So one bank could have both types of bank accounts like asset and liability
 	# So cr_or_dr should be judged only on basis of withdrawal and deposit and not account type
 	cr_or_dr = "credit" if transaction.withdrawal > 0 else "debit"
+	# filter_by_date = f"AND je.posting_date between '{from_date}' and '{to_date}'"
+	order_by = " je.posting_date"
+	if filtered_by_reference_date == "1":
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]: Here too

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]

**Comment by sonali8848** [comment]
OKK , I will do

**Comment by deepeshgarg007** [comment]
@sonali8848 Please also add in the description why these changes were needed, what changes are added is well understood but why they were needed is still a bit confusing to me.

Unit tests also failing after the changes

**Comment by sonali8848** [comment]
Okk, I will add the proper description, but unit testing is failing because
we have passed the filters in get_linked_payments function, in  test cases
the filters are not passed that's why it is failing.

**Comment by deepeshgarg007** [comment]
> Okk, I will add the proper description, but unit testing is failing because we have passed the filters in get_linked_payments function, in test cases the filters are not passed that's why it is failing

So pass the filters from the backend, is there any issue you are facing with this?

**Comment by sonali8848** [comment]
I will add the filters on test cases in test_bank_transaction

**Comment by sonali8848** [comment]
Hi Deepesh,

The failed test cases are not linked with bank reconciliation, which we
have done. The link is mentioned below. Kindly check.
https://github.com/frappe/erpnext/actions/runs/3805772367/jobs/6498112341

**Comment by deepeshgarg007** [comment]
> @sonali8848 Please also add in the description why these changes were needed, what changes are added is well understood but why they were needed is still a bit confusing to me.

@sonali8848 this is still unanswered, can you update this as well

**Comment by sonali8848** [comment]
Hi Deepesh,

I have already added it in the comment section. The link is mentioned
below. Kindly check.
https://github.com/frappe/erpnext/pull/33271#issue-1484503677

Thanks

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js [hunk 10]
```diff
@@ -25,6 +25,24 @@ frappe.ui.form.on("Bank Reconciliation Tool", {
 		frappe.require("bank-reconciliation-tool.bundle.js", () =>
 			frm.trigger("make_reconciliation_tool")
 		);
+		frm.add_custom_button(__('Auto Reconcile'), function(){
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js]:
```suggestion
		frm.add_custom_button(__('Auto Reconcile'), function() {
```

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js [hunk 11]
```diff
@@ -25,6 +25,24 @@ frappe.ui.form.on("Bank Reconciliation Tool", {
...
+			frappe.call({
+				method:"erpnext.accounts.doctype.bank_reconciliation_tool.bank_reconciliation_tool.auto_reconcile_vouchers",
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js]:
```suggestion
				method: "erpnext.accounts.doctype.bank_reconciliation_tool.bank_reconciliation_tool.auto_reconcile_vouchers",
```

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js [hunk 12]
```diff
@@ -25,6 +25,24 @@ frappe.ui.form.on("Bank Reconciliation Tool", {
...
+				args: {
+					bank_account:frm.doc.bank_account,
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js]:
```suggestion
					bank_account: frm.doc.bank_account,
```

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js [hunk 13]
```diff
@@ -25,6 +25,24 @@ frappe.ui.form.on("Bank Reconciliation Tool", {
...
+					from_date: frm.doc.bank_statement_from_date,
+					to_date: frm.doc.bank_statement_to_date,
+					filtered_by_reference_date:frm.doc.filtered_by_reference_date,
+					from_reference_date:frm.doc.from_reference_date,
+					to_reference_date:frm.doc.to_reference_date,
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.js]:
```suggestion
					filtered_by_reference_date: frm.doc.filtered_by_reference_date,
					from_reference_date: frm.doc.from_reference_date,
					to_reference_date: frm.doc.to_reference_date,
```

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 14]
```diff
@@ -265,6 +266,82 @@ def create_payment_entry_bts(
...
+		for r in linked_payments:
+			vouchers.append(
+				{
+					"payment_doctype": r[1],
+					"payment_name": r[2],
+					"amount": r[4],
+				}
+			)
+		# vouchers = frappe.as_json(voucherss)
+		# vouchers = json.loads(voucherss)
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]: Remove comments

File: erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py [hunk 15]
```diff
@@ -265,6 +266,82 @@ def create_payment_entry_bts(
...
+					"payment_entry": voucher["payment_name"],
+					"allocated_amount": allocated_amount,
+				},
+			)
+			matched_transaction.append(str(transaction.name))
+		transaction.save()
+		transaction.update_allocations()
+	matched_transaction_len = len(set(matched_transaction))
+	if matched_transaction_len == 0:
+		frappe.msgprint(_("There is no any matched reference number for reconciliation"))
```
Comment (deepeshgarg007) [inline erpnext/accounts/doctype/bank_reconciliation_tool/bank_reconciliation_tool.py]:
```suggestion
		frappe.msgprint(_("No matching references found for auto reconciliation"))
```

File: erpnext/public/js/bank_reconciliation_tool/data_table_manager.js [hunk 16]
```diff
@@ -17,6 +22,8 @@ erpnext.accounts.bank_reconciliation.DataTableManager = class DataTableManager {
 				"erpnext.accounts.doctype.bank_reconciliation_tool.bank_reconciliation_tool.get_bank_transactions",
 			args: {
 				bank_account: this.bank_account,
+				from_date:this.bank_statement_from_date,
+				to_date:this.bank_statement_to_date
```
Comment (deepeshgarg007) [inline erpnext/public/js/bank_reconciliation_tool/data_table_manager.js]:
```suggestion
				from_date: this.bank_statement_from_date,
				to_date: this.bank_statement_to_date
```

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]
Just some code cleanup changes, please add a simple test case for the Auto Reconciliation use case and also a working GIF in the description

**Pull request merged**